src/ffi/ios/structs.rs

```diff
@@ -0,0 +1,177 @@
+    Get,
+}
+
+impl From<Method> for http::Method {
+    fn from(method: Method) -> Self {
+        match method {
+            Method::Get => http::Method::GET,
+        }
+    }
+}
+
+#[repr(C)]
+pub struct Request {
+    method: Method,
+    url: CFURLRef,
+    /// CFDictionary<CFString, CFArray<CFData>>,
+    headers: CFDictionaryRef,
+    body: CFDataRef,
+}
+
+impl TryFrom<Request> for super::Request {
+    type Error = anyhow::Error;
+
+    fn try_from(request: Request) -> anyhow::Result<Self> {
+        let url_ios = unsafe { CFURL::wrap_under_get_rule(request.url) };
+        let url_raw = url_ios.get_string();
+        let url_raw: Cow<_> = (&url_raw).into();
+        let url: Uri = url_raw.parse().context("parse url")?;
+
+        let headers_ios = unsafe {
+            CFDictionary::<CFString, CFArray<CFData>>::wrap_under_get_rule(request.headers)
+        };
+        let body_ios = unsafe { CFData::wrap_under_get_rule(request.body) };
+
+        let mut ret = http::Request::builder()
+            .method::<http::Method>(request.method.into())
+            .header("Host", url.host().context("no host in request")?)
+            .version(http::Version::HTTP_10)
+            .uri(url)
+            .body(body_ios.bytes().to_vec())
+            .context("invalid request")?;
+
+        let (headers_keys, headers_values) = headers_ios.get_keys_and_values();
+        let hm = ret.headers_mut();
+        for (key, values) in headers_keys
+            .into_iter()
+            .map(|k| unsafe { CFString::from_void(k) })
+            .zip(
+                headers_values
+                    .into_iter()
+                    .map(|v| unsafe { CFArray::<CFData>::from_void(v) }),
+            )
+        {
+            let key_raw: Cow<_> = (&(*key)).into();
+            let key = HeaderName::try_from(key_raw.as_ref()).context("invalid header name")?;
+
+            for value in values.iter() {
+                let value = HeaderValue::try_from(value.bytes()).context("invalid header value")?;
+                hm.append(&key, value);
+            }
+        }
+
+        Ok(super::Request(ret))
+    }
+}
+
+#[repr(C)]
+pub struct Response {
+    status: u16,
+    /// CFDictionary<CFString, CFArray<CFData>>,
+    headers: CFDictionaryRef,
+    body: CFDataRef,
+}
+
+impl TryFrom<super::Response> for Response {
+    type Error = anyhow::Error;
+
+    fn try_from(resp: super::Response) -> anyhow::Result<Self> {
+        let (parts, body) = resp.0.into_parts();
+
+        let mut headers_with_vec = HashMap::with_capacity(parts.headers.keys_len());
+        let mut last_header_name = None;
+        for (k, v) in parts.headers {
+            let k = k
+                .map(|k| k.as_str().to_owned())
+                .or(last_header_name)
+                .context("no header name")?;
+            last_header_name = Some(k.clone());
+
+            headers_with_vec.entry(k).or_insert_with(Vec::new).push(v);
+        }
+
+        // TODO avoid copies
```

Comment:
Is this important?

Reply:
Nope, it was hard to get it working in the first place, not gonna find a new trick